processor/swok8sworkloadtypeprocessor/processor.go

```diff
@@ -0,0 +1,224 @@
+	if attrVal, ok := attributes.Get(attrName); ok && attrVal.Type() == pcommon.ValueTypeStr {
+		return attrVal.Str()
+	}
+	return ""
+}
+
+func (cp *swok8sworkloadtypeProcessor) setWorkloadTypeForGivenDatapointAttributes(attributes pcommon.Map) error {
+	for _, workloadMapping := range cp.config.WorkloadMappings {
+		name := cp.getAttribute(attributes, workloadMapping.NameAttr)
+		if name == "" {
+			continue
+		}
+		namespace := cp.getAttribute(attributes, workloadMapping.NamespaceAttr)
+		var workloadKey string
+		if namespace != "" {
+			workloadKey = fmt.Sprintf("%s/%s", namespace, name)
+		} else {
+			workloadKey = name
+		}
+
+		for _, workloadType := range workloadMapping.ExpectedTypes {
+			workload, exists, err := cp.informers[workloadType].GetStore().GetByKey(workloadKey)
+			if err != nil {
+				return fmt.Errorf("error when retrieving workload '%s' from cache: %w", workloadKey, err)
+			}
+			if exists {
+				workloadObject, ok := workload.(runtime.Object)
+				if !ok {
+					return fmt.Errorf("unexpected type '%T' for workload '%s'", workload, workloadKey)
+				}
+				kind := workloadObject.GetObjectKind().GroupVersionKind().Kind
+				if kind == "" {
+					return fmt.Errorf("workload '%s' has no kind", workloadKey)
+				}
+				attributes.PutStr(workloadMapping.WorkloadTypeAttr, kind)
+				break
+			}
+		}
+	}
+
+	return nil
+}
+
+func (cp *swok8sworkloadtypeProcessor) processMetrics(ctx context.Context, md pmetric.Metrics) (pmetric.Metrics, error) {
+	resourceMetrics := md.ResourceMetrics()
+	for rmIdx := range resourceMetrics.Len() {
+		rm := resourceMetrics.At(rmIdx)
+		scopeMetrics := rm.ScopeMetrics()
+		for smIdx := range scopeMetrics.Len() {
+			sm := scopeMetrics.At(smIdx)
+			metrics := sm.Metrics()
+			for mIdx := range metrics.Len() {
+				m := metrics.At(mIdx)
+				switch m.Type() {
+				case pmetric.MetricTypeGauge:
+					datapoints := m.Gauge().DataPoints()
+					for dpIdx := range datapoints.Len() {
+						dp := datapoints.At(dpIdx)
+						err := cp.setWorkloadTypeForGivenDatapointAttributes(dp.Attributes())
+						if err != nil {
+							return pmetric.Metrics{}, fmt.Errorf("error getting workload type for metric '%s': %w", m.Name(), err)
+						}
+					}
+				case pmetric.MetricTypeSum:
+					datapoints := m.Sum().DataPoints()
+					for dpIdx := range datapoints.Len() {
+						dp := datapoints.At(dpIdx)
+						err := cp.setWorkloadTypeForGivenDatapointAttributes(dp.Attributes())
+						if err != nil {
+							return pmetric.Metrics{}, fmt.Errorf("error getting workload type for metric '%s': %w", m.Name(), err)
+						}
+					}
+				case pmetric.MetricTypeHistogram:
+					datapoints := m.Histogram().DataPoints()
+					for dpIdx := range datapoints.Len() {
+						dp := datapoints.At(dpIdx)
+						err := cp.setWorkloadTypeForGivenDatapointAttributes(dp.Attributes())
+						if err != nil {
+							return pmetric.Metrics{}, fmt.Errorf("error getting workload type for metric '%s': %w", m.Name(), err)
+						}
+					}
```

Comment:
This block is pretty much duplicated through all metric types, can you extract that to functions to avoid this duplication?

Reply:
After upgrading to OTEL collector 0.123.0 and introducing some generics here, the code should be much shorter now.